overlord/servicestate/quota_handlers.go

```diff
@@ -781,31 +803,39 @@ func ensureSnapServicesStateForGroup(st *state.State, grp *quota.Group, opts *en
 	return restartSnapServices(st, appsToRestartBySnap)
 }
 
-func ensureGroupIsNotMixed(group string, allGrps map[string]*quota.Group) error {
-	grp, ok := allGrps[group]
-	if ok && len(grp.SubGroups) != 0 && len(grp.Snaps) != 0 {
-		return fmt.Errorf("quota group %q has mixed snaps and sub-groups, which is no longer supported; removal and re-creation is necessary to modify it", group)
+// groupEnsureOnlySnapsOrServices is a helper which returns an error if the snaps and/or
+// services we are trying to add would result in a mixed group. A group is considered
+// mixed if:
+// 1. The group would end up containing both snaps and services
+// 2. The group would end up containing both services and groups
+func groupEnsureOnlySnapsOrServices(snapsToAdd, servicesToAdd []string, grp *quota.Group) error {
+	if len(snapsToAdd) > 0 && len(servicesToAdd) > 0 {
+		return fmt.Errorf("cannot mix services and snaps in the same quota group")
+	}
+
+	if (len(grp.Services) > 0 && len(snapsToAdd) > 0) || (len(grp.Snaps) > 0 && len(servicesToAdd) > 0) {
+		return fmt.Errorf("cannot mix services and snaps in the same quota group")
 	}
 	return nil
 }
 
-// ensureGroupHasNoSubgroups returns true if the group has no sub-groups or it doesn't exist,
-// otherwise turns false.
-func ensureGroupHasNoSubgroups(group string, allGrps map[string]*quota.Group) bool {
-	grp, ok := allGrps[group]
-	if ok && len(grp.SubGroups) != 0 {
-		return false
+// ensureParentIsNotMixed returns true if the parent is not a mixed group. A group
+// is considered mixed if it contains both snaps and sub-groups. In this case we only
+// allow services to be added to sub-groups, as we do not support nesting of snaps.
+func ensureParentIsNotMixed(grp *quota.Group) bool {
```

Comment:
since this returns a boolean, I'd call it `groupIsMixed()` and invert the return value. Also, I'd remove the "parent" from the name and the comments, because otherwise it might make look like that this function is checking the parent of the given group.

Sorry for not noticing this earlier!

Reply:
Really good feedback, when I'm deep into solving something these kind of names make so much more sense than when you step away from your code and come back later. I agree something was weird here!